File: Difficulty: Medium/Construct a Full Binary Tree/construct-a-full-binary-tree.cpp
//{ Driver Code Starts
//Initial Template for C++
#include<bits/stdc++.h>
using namespace std;

struct Node
{
	int data;
	struct Node *left;
	struct Node *right;
	
	Node(int x){
	    data = x;
	    left = NULL;
	    right = NULL;
	}
};


// } Driver Code Ends
/* Structre of the Node of the Binary Tree is as follows
struct Node
{
        int data;
        struct Node *left, *right;
};
*/
// your task is to complete this function
// function should create a new binary tree
// function should return the root node to the 
// new binary tree formed
/* Structure of the Node of the Binary Tree is as follows
struct Node
{
        int data;
        struct Node *left, *right;
};
*/

class Solution{
  public:
  // Helper function to construct the binary tree
  Node* help(int &index, int start, int end, int pre[], int preMirror[], int n){
        // Base case: if start is greater than end or if we've processed all nodes
        if (start > end || index >= n) {
            return NULL;
        }
        
        // If we reach the last node or a single node in the range, create and return a leaf node
        if (index == n - 1 || start == end) {
            Node* temp = new Node(pre[index]);
            index++;
            return temp;
        }

        // Create a new node with the current value in pre[] at the current index
        Node* root = new Node(pre[index]);
        index++;  // Move to the next element in pre[]
        
        // Search for the value of pre[index] in the preMirror[] array to determine the left and right subtree
        for (int i = start; i <= end; i++) {
            if (pre[index] == preMirror[i]) {
                // Recursively build the left subtree
                root->left = help(index, i, end, pre, preMirror, n);
                
                // Recursively build the right subtree
                root->right = help(index, start + 1, i - 1, pre, preMirror, n);
                break;  // Stop once we've found the appropriate split
            }
        }
        
        return root;  // Return the root node
    }

    // Main function to construct the binary tree from pre-order and pre-mirror arrays
    Node* constructBinaryTree(int pre[], int preMirror[], int size)
    {
        int index = 0;  // Initialize index for pre[] array
        return help(index, 0, size - 1, pre, preMirror, size);  // Call helper function to construct tree
    }
};


//{ Driver Code Starts.

void printInorder(Node* node)
{
	if (node == NULL)return;
	printInorder(node->left);
	cout<<node->data<<" ";
	printInorder(node->right);
}

// Driver program to test above functions
int main()
{
    int t;
    cin>>t;
    while(t--)
    {
        int n;
        cin>>n;
        int preOrder[n];
	    int preOrderMirror[n];
	    for(int i=0; i<n; i++)cin>>preOrder[i];
	    for(int i=0; i<n; i++)cin>>preOrderMirror[i];
	    Solution obj;
	    printInorder(obj.constructBinaryTree(preOrder, preOrderMirror, n));
	    cout<<endl;
    
cout << "~" << "\n";
}
	return 0;
}
// } Driver Code Ends

File: Difficulty: Medium/Construct a Full Binary Tree/construct-a-full-binary-tree.js
//{ Driver Code Starts
//Initial Template for javascript

'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data',  inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', _ => {
    inputString = inputString.trim().split('\n').map(string => {
        return string.trim();
    });
    
    main();    
});

function readLine() {
    return inputString[currentLine++];
}

class Node{
    constructor(data){
        this.data = data;
        this.left = null;
        this.right = null;
    }
}

function buildTree(str){
   // Corner Case
   if(str.length === 0 || str[0] === "N")
       return null;

   // Create the root of the tree
   let root = new Node(parseInt(str[0]));

   // Push the root to the queue
   let queue = [];
   let start = 0;
   queue.push(root);

   // Starting from the second element
   let i = 1;
   while(queue.length !== start && i < str.length) {

       // Get and remove the front of the queue
       let currNode = queue[start];
       start++;

       // Get the current node's value from the string
       let currVal = str[i];

       // If the left child is not null
       if(currVal !== "N") {

           // Create the left child for the current node
           currNode.left = new Node(parseInt(currVal));

           // Push it to the queue
           queue.push(currNode.left);
       }

       // For the right child
       i++;
       if(i >= str.length)
           break;
       currVal = str[i];

       // If the right child is not null
       if(currVal !== "N") {

           // Create the right child for the current node
           currNode.right = new Node(parseInt(currVal));

           // Push it to the queue
           queue.push(currNode.right);
       }
       i++;
   }

   return root;
}

let s = "";
function inorder(root){
  if(root === null) return;
  inorder(root.left);
  s = s + root.data + " ";
  inorder(root.right);
}
 

function main() {
    let t = parseInt(readLine());
    let i = 0;
    for(;i<t;i++)
    {
        let size = parseInt(readLine());
        let preorder = readLine().trim().split(" ").map((x) => parseInt(x));
        let preorderMirror = readLine().trim().split(" ").map((x) => parseInt(x));
        let obj = new Solution();
        let root = obj.constructBinaryTree(preorder,preorderMirror,size);
        s="";
        inorder(root);
        console.log(s);  
    }
}
// } Driver Code Ends


// User function Template for javascript

/*
class Node{
    constructor(data){
        this.data = data;
        this.left = null;
        this.right = null;
    }
}
*/

/**
 * @param {number[]} preorder
 * @param {number[]} preorderMirror
 * @param {number} size
 * @return {Node}
 */
class Solution {
  constructBinaryTree(preorder, preorderMirror, size) {
    // Create postorder by reversing preorderMirror
    let postorder = preorderMirror.reverse();
    
    // Create a map to store the indices of elements in postorder
    let map = {};
    for (let i = 0; i < postorder.length; ++i) {
        map[postorder[i]] = i;
    }
    
    // Helper function to construct the tree
    function helper(preStart, preEnd, postStart, postEnd) {
        // If the range is invalid, return null (base case)
        if (preStart > preEnd) return null;
        
        // Create a root node with the current value in preorder
        let root = new Node(preorder[preStart]);
        
        // If the range is a single element, return the node
        if (preStart === preEnd) return root;
        
        // Find the index of the left child in postorder and calculate the offset
        let idx = map[preorder[preStart + 1]];
        let offset = idx - postStart + 1;
        
        // Recursively construct left and right subtrees
        root.left = helper(preStart + 1, preStart + offset, postStart, idx);
        root.right = helper(preStart + offset + 1, preEnd, idx + 1, postEnd - 1);
        
        return root;
    }
    
    // Call the helper function to construct the tree
    return helper(0, preorder.length - 1, 0, postorder.length - 1);
  }
}
